src/monty/collections.py

```diff
@@ -20,126 +30,202 @@ def tree() -> collections.defaultdict:
 
     Usage:
         x = tree()
-        x['a']['b']['c'] = 1
+        x["a"]["b"]["c"] = 1
 
     Returns:
         A tree.
     """
     return collections.defaultdict(tree)
 
 
-class frozendict(dict):
+class ControlledDict(collections.UserDict, ABC):
     """
-    A dictionary that does not permit changes. The naming
-    violates PEP8 to be consistent with standard Python's "frozenset" naming.
+    A base dictionary class with configurable mutability.
+
+    Attributes:
+        _allow_add (bool): Whether new keys can be added.
+        _allow_del (bool): Whether existing keys can be deleted.
+        _allow_update (bool): Whether existing keys can be updated.
+
+    Configurable Operations:
+        This class allows controlling the following dict operations (refer to
+        https://docs.python.org/3.13/library/stdtypes.html#mapping-types-dict for details):
+
+        - Adding or updating items:
+            - setter method: `__setitem__`
+            - `setdefault`
+            - `update`
+
+        - Deleting items:
+            - `del dict[key]`
+            - `pop(key)`
+            - `popitem`
+            - `clear()`
     """
 
-    def __init__(self, *args, **kwargs) -> None:
-        """
-        Args:
-            args: Passthrough arguments for standard dict.
-            kwargs: Passthrough keyword arguments for standard dict.
-        """
-        dict.__init__(self, *args, **kwargs)
+    _allow_add: bool = True
+    _allow_del: bool = True
+    _allow_update: bool = True
 
-    def __setitem__(self, key: Any, val: Any) -> None:
-        raise KeyError(f"Cannot overwrite existing key: {str(key)}")
+    def __init__(self, *args, **kwargs) -> None:
+        """Temporarily allow add during initialization."""
+        original_allow_add = self._allow_add
 
-    def update(self, *args, **kwargs) -> None:
-        """
-        Args:
-            args: Passthrough arguments for standard dict.
-            kwargs: Passthrough keyword arguments for standard dict.
-        """
-        raise KeyError(f"Cannot update a {self.__class__.__name__}")
+        try:
+            self._allow_add = True
+            super().__init__(*args, **kwargs)
+        finally:
+            self._allow_add = original_allow_add
 
+    # Override add/update operations
+    def __setitem__(self, key, value) -> None:
+        """Forbid adding or updating keys based on _allow_add and _allow_update."""
+        if key not in self.data and not self._allow_add:
+            raise TypeError(f"Cannot add new key {key!r}, because add is disabled.")
+        elif key in self.data and not self._allow_update:
+            raise TypeError(f"Cannot update key {key!r}, because update is disabled.")
 
-class Namespace(dict):
```

Comment:
If I understand correctly, `Namespace` is a dict that only allows adding new items but not allow updating existing items, however I don't really understand why it's called `Namespace`?